HYOK docs (#645)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/hold-your-own-key/configure.mdx

---
page_title: Configure and manage keys
description: >-
  Learn how to set up hold your own key encryption for HCP Terraform, letting you secure your sensitive data in state and plan files using a key from an external key management system.
tfc_only: true
---

# Configure and manage keys

Configure your key management system (KMS) and HCP Terraform to use the hold your own key (HYOK) feature to encrypt your sensitive Terraform artifacts.

## Background

HCP Terraform lets you configure multiple HYOK configurations for an
organization. Each HYOK configuration corresponds to a single key in a
KMS. Additional HYOK configurations can point to different keys in the same KMS, or keys in others KMSs.

> When you set up HYOK, you will designate one configuration as your organization's **primary** configuration. When you enable HYOK on a workspace, HCP Terraform automatically secures that workspace's artifacts using your organization's primary HYOK configuration.

Reviewer: ```suggestion
When you set up HYOK, you designate one configuration as your organization's **primary** configuration. When you enable HYOK on a workspace, HCP Terraform automatically secures that workspace's artifacts using your organization's primary HYOK configuration.
```

```json
{
  "Version": "2012-10-17",
  "Statement": [
      {
          "Effect": "Allow",
          "Principal": {
              "Federated": "<the ARN of your new Identity Provider>"
          },
          "Action": "sts:AssumeRoleWithWebIdentity",
          "Condition": {
              "StringEquals": {
                  "app.terraform.io:aud": "aws.workload.identity"
              },
              "StringLike": {
                  "app.terraform.io:sub": "organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>"

              }
          }
      }
  ]
}
```

The **Condition** field matches against the subject claim of HCP Terraform's
JWT. For HYOK, the subject claim is: `organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>`.

Wildcards are allowed when matching claims. For example, if you want to use the
same role for all keys within your HCP Terraform organization, you can configure
your subject claim to match all HYOK configurations in your organization:
`organization:<ORG_NAME>:hyok_config:*`.

For additional security we recommend that you choose a configuration name now,
so that only the specified HYOK configuration may assume this role:

> `organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>`. The hyok configuration

Reviewer: ```suggestion
`organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>`. The HYOK configuration
```

---
page_title: Configure and manage keys
description: >-
  Learn how to set up hold your own key encryption for HCP Terraform, letting you secure your sensitive data in state and plan files using a key from an external key management system.
tfc_only: true
---

# Configure and manage keys

Configure your KMS and HCP Terraform before you use hold your own key (HYOK) to
encrypt your sensitive artifacts.

## Background

HCP Terraform allows you to configure multiple HYOK configurations in your
organization settings. Each HYOK configuration corresponds to a single key in a
KMS. Other HYOK configurations could point to different keys in the same KMS, or
keys in multiple KMSs.

A single HYOK configuration is marked as your organization's **primary**
configuration. When you enable HYOK on a workspace, its artifacts are
automatically secured with the key specified in the primary configuration.

## Requirements

To get started using the Hold Your Own Key (HYOK) feature, there are a few steps
you will need to take prior to creating your first HYOK configuration and
enabling HYOK for your workspaces.

> - Ensure your organization is on the Premium Tier, as HYOK is only supported for

Reviewer: Great point! We'll actually replace this with a standardized call out partial up top! Leaving this comment to remind me to actually do that 😂

> - Ensure your organization is on the Premium Tier, as HYOK is only supported for

Reviewer: (Resolving because I did this!)

  Premium Tier organizations.
- Ensure your organization has at least one agent pool hosted with request

>   forwarding enabled. The minimum agent version for HYOK is 1.23.0.

Reviewer: Added!

name>`. The hyok configuration name can contain letters, numbers, hyphens, and
underscores.

</Tab>
</Tabs>

## Configure HYOK in HCP Terraform

Once your KMS has been configured with a key encryption key the necessary OIDC
trust relationship to HCP Terraform, the next step is to create an **HYOK
configuration** for your HCP Terraform organization.

Each HYOK configuration in your organization represents a single KMS key
encryption key with potentially multiple key versions. You can configure
multiple keys in different HYOK configurations, across one or many different
KMSs.

Navigate to **Organization Settings** then **HYOK Encryption**. Open the HYOK
configuration wizard by clicking the **Configure a key** button.

### Choose agent pool

On the first step of the wizard, select an agent pool that is hosting agents
with [Request Forwarding
enabled](/terraform/cloud-docs/agents/request-forwarding).

HCP Terraform uses this agent pool to:

- Generate and encrypt data encryption keys for workspaces onboarded with HYOK.
- Re-encrypt data keys when an HYOK configuration is revoked.
- Connect to your KMS to check for new key versions.

> It is important that the agent pool contains agents with request forwarding

Reviewer: Added!

If you are using Vault Enterprise with a custom CA, enter the TLS CA Certificate
which authenticates it.

</Tab>
</Tabs>

### Configure key settings

On the final page of the wizard, tell HCP Terraform which key to use for this
HYOK configuration.

Provide the configuration with a unique name to identify it within HCP Terraform. Your configuration name can be any combination of letters, numbers, hyphens, and underscores, and is included in the subject claim of HYOK JWTs issued by HCP Terraform. So if you configured your KMS to match against your hyok_config name, make sure to use the same name for your HYOK configuration..

<Tabs>
<Tab heading="Azure" group="azure">

To configure your key in HCP Terraform, you will need the following values from Vault:
The full Key URI for the key you created in the previous step.
This is in the format https://<key vault name>.vault.azure.net/keys/<key name>

</Tab>
<Tab heading="AWS" group="aws">

To configure your key in HCP Terraform, you will need the following values from AWS:
The full Key ARN for the KMS key you created in the previous step.
The Region that the key is located in.

</Tab>
<Tab heading="GCP" group="gcp">

> The Key Ring Name for the key ring you created in the previous step.

Reviewer: (Made these bullet points)

The Key Name for the specific key in that ring.
The Location that the key ring was created in.
E.g. global

</Tab>
<Tab heading="Vault" group="vault">

The Name of the key you created in the Transit secrets engine.
In the Vault configuration example above, this was unique-prod

</Tab>
</Tabs>

## Test the key configuration

> To test your HYOK configuration, click Test Connection on the last step of the
> Configure a new key wizard during set up. Clicking this button causes HCP

Reviewer: We'll actually do this with bold! We use bold for UI elements in the docs:
```suggestion
To test your HYOK configuration, click **Test Connection** on the last step of the
**Configure a new key** wizard during set up. Clicking **Test Connection** tells HCP
```

HCP Terraform, it migrates all of the workspaces that used the revoked
configuration to use the primary configuration.

To manually promote a configuration to primary, navigate to your organization's
Settings, and then to HYOK Encryption. Click Promote to primary on the desired
configuration. Promotion to primary will also happen automatically for the first
HYOK configuration with a status of "Available".

Once you’ve configured HCP Terraform to use your key, select a workspace to
encrypt data for by enabling HYOK for that workspace. Please note that once HYOK
is enabled for a workspace, it cannot be reversed. Once you enable HYOK on a
given workspace, HCP Terraform will generate a data encryption key, connect to
your KMS to encrypt it with your key encryption key, and store the encrypted
DEK.

To enable HYOK for a workspace, go to a workspace’s settings, and navigate to
HYOK Encryption. Check Enable HYOK for this workspace to enable HYOK and encrypt
data for the state and plan files in your workspace.

Any future runs on this workspace now produce encrypted state and plan files.
You can test this by performing a run. State versions are annotated with the

> name of the HYOK configuration securing them:

Reviewer 2: ```suggestion
name of the HYOK configuration securing them.
```

## Revoke a key configuration

HCP Terraform keeps track of which HYOK configurations are used to secure each
of your workspace’s artifacts. If you wish to stop using a given KEK or a
specific key version, you may revoke it within HCP Terraform. For example, if a
key version is too old or has been leaked, then revoking it in HCP Terraform
instructs the platform to stop using that key for encryption, and to migrate all
existing data off of it.

> 

Reviewer: ```suggestion

An important distinction to make is that HCP Terraform does not manage keys within your KMS. Meaning, that revoking a key configuration in HCP Terraform does not revoke that key within your KMS. Revoking a key configuration in HCP Terraform lets you migrate data off of the key, letting you can then revoke and safely delete the corresponding key from your KMS.

```

When a key version is revoked within HCP Terraform, HCP Terraform automatically
re-secures all of the artifacts it currently secures with the latest version of
the same key. When an entire HYOK configuration is revoked, HCP Terraform
re-secures the artifacts with the primary configuration.

When a key or key version is revoked, HCP Terraform does not need to directly
re-encrypt all the artifacts it secures. Instead, HCP Terraform uses an HCP
Terraform agent to re-encrypt all of the DEKs that are encrypted with that key.
This makes the re-encryption process much quicker and requires less transferring
data than re-encrypting the artifacts themselves.

> HCP Terraform cannot manage the keys within your KMS: revoking a key in HCP

Reviewer: I agree we need to call more attention to this! I think we should just move this info to the very top of this section to give it the attention it deserves!
 
 **For background context:** 
* we only bold UI elements in our documentation. 
* We could alternatively put the information in a callout, but we also recommend against callouts in general because they disrupt the reading experience. So, I generally just move important stuff to the top of a section 😂

For additional security we recommend that you choose a configuration name now,
so that only the specified HYOK configuration may assume this role:

> `organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>`. The hyok configuration

Reviewer: ```suggestion
`organization:<ORG_NAME>:hyok_config:<HYOK_CONFIG_NAME>`. The HYOK configuration
```
